### PR TITLE
[enhancement]RayDashboardClient deal with abnormal http status

### PR DESCRIPTION
**Pull request by kingeasternsun**
## Why are these changes needed?

in `dashboard_httpclient`,  dealing with the abnormal http status  before Unmarshal will make the error more precise。

AND  taking this https://github.com/ray-project/kuberay/issues/1453 in consideration,  make 404 returned by `GetJobInfo`  as  `ErrJobInfoNotFound`

## Related issue number

https://github.com/ray-project/kuberay/issues/1453

## Checks

- [x] I've made sure the tests are passing. 
- Testing Strategy
   - [x] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

image

### CODE REVIEW

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go [hunk 1]
```diff
@@ -352,14 +357,18 @@ func (r *RayDashboardClient) GetJobInfo(ctx context.Context, jobId string) (*Ray
 	defer resp.Body.Close()
 
 	if resp.StatusCode == http.StatusNotFound {
-		return nil, nil
+		return nil, ErrJobInfoNotFound
```
Comment (rueian) [inline ray-operator/controllers/ray/utils/dashboard_httpclient.go:360]: Hi @kingeasternsun,

I noticed that this `if resp.StatusCode == http.StatusNotFound` check should be moved to after the below `io.ReadAll(resp.Body)`

Could you also help correct that? Thanks!

**Comment by rueian** [comment]
And here:

https://github.com/ray-project/kuberay/blob/9425e7fe39c2eac7c857d218cd3e106fdd02cd5c/ray-operator/controllers/ray/rayjob_controller.go#L199-L206

We use the `GetJobInfo` for checking dashboard readiness and expect it to return a 404, which was previously represented as `nil`. Therefore, the `err != nil` should be changed to `err != nil && !error.Is(err, ErrJobInfoNotFound)`.

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go (hunk 1 above)
Comment (kingeasternsun) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go:360]: Thanks for your review, I will fix that.

Comment (kingeasternsun) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go:360]: @rueian    this place  `io.ReadAll(resp.Body)`  should I check the return error as before, or  changed to  `body, _ := io.ReadAll(resp.Body)` like other methods below?

Comment (rueian) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go]: I think the current implementation is ok. No need to change.

**Review by rueian — APPROVED** [review]

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go (hunk 1 above)
Comment (kingeasternsun) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go]: OK

**Comment by kingeasternsun** [comment]
@rueian  thanks for your approving,  But this PR is still Open, Is there anything wrong?

**Comment by rueian** [comment]
Hi @kingeasternsun,

I approved your PR on the behalf of the original issue creator. You would still need others approvals.

**Comment by kingeasternsun** [comment]
@kevin85421  Hi, will u please review this ?

File: ray-operator/controllers/ray/rayjob_controller.go [hunk 2]
```diff
@@ -197,10 +198,10 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
 			return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
 		}
 		// Check the dashboard readiness by checking the err return from rayDashboardClient.GetJobInfo.
-		// Note that rayDashboardClient.GetJobInfo returns no error in the case of http.StatusNotFound.
+		// Note that rayDashboardClient.GetJobInfo returns ErrJobInfoNotFound in the case of http.StatusNotFound.
 		// This check is a workaround for https://github.com/ray-project/kuberay/issues/1381.
 		rayDashboardClient.InitClient(clientURL)
-		if _, err = rayDashboardClient.GetJobInfo(ctx, rayJobInstance.Status.JobId); err != nil {
+		if _, err = rayDashboardClient.GetJobInfo(ctx, rayJobInstance.Status.JobId); err != nil && !goerrors.Is(err, utils.ErrJobInfoNotFound) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller.go:204]: This is a bit trivial. We use the API to check for readiness, but we still need to determine which error it is. We may need a new API to check the readiness in the future.

Comment (kevin85421) [reply ray-operator/controllers/ray/rayjob_controller.go:204]: Although I hope to reduce the interaction between KubeRay and the Ray head as much as possible, the check is still a must at this moment. @architkulkarni is there any API endpoint to check the readiness of the Ray dashboard?

**Review by kevin85421 — COMMENTED** [review]
Thank you for the contribution! I am not entirely confident that the mock test replicates Ray's behavior accurately. Would you mind designing a manual test to verify it in an actual K8s cluster? @architkulkarni may have the context about the Ray behavior.

**Comment by kevin85421** [comment]
@kingeasternsun would you mind fixing the lint error in CI? Thanks!

File: ray-operator/controllers/ray/rayjob_controller.go (hunk 2 above)
Comment (architkulkarni) [reply ray-operator/controllers/ray/rayjob_controller.go:204]: I don't think there's any special API for this. I think I've seen `/api/version` https://docs.ray.io/en/latest/cluster/running-applications/job-submission/api.html#/paths/~1api~1version/get used for this purpose.

**Comment by architkulkarni** [comment]
> Thank you for the contribution! I am not entirely confident that the mock test replicates Ray's behavior accurately. Would you mind designing a manual test to verify it in an actual K8s cluster? @architkulkarni may have the context about the Ray behavior.

@kevin85421 Hmm, what behavior are you interested in here? Don't the existing RayJob e2e tests cover this?

If we change the "dashboard health check" to use `/api/version` would that be preferable? That way we don't have to introduce logic to deal with the 404 from GetJobInfo and worry about how to test that logic

**Comment by kevin85421** [comment]
> @kevin85421 Hmm, what behavior are you interested in here? Don't the existing RayJob e2e tests cover this?

I am wondering what the Ray dashboard will return if users look up a non-existent JobID. Will it return a 404? I am not sure whether the mock tests have the same behavior as the Ray dashboard or not.

> If we change the "dashboard health check" to use `/api/version` would that be preferable? That way we don't have to introduce logic to deal with the 404 from GetJobInfo and worry about how to test that logic

Does successfully retrieving `/api/version` from the dashboard indicate that it is ready for job requests? If so, this would be a better solution.

**Comment by architkulkarni** [comment]
> > @kevin85421 Hmm, what behavior are you interested in here? Don't the existing RayJob e2e tests cover this?
> 
> I am wondering what the Ray dashboard will return if users look up a non-existent JobID. Will it return a 404? I am not sure whether the mock tests have the same behavior as the Ray dashboard or not.

Ah I see. Yes, that is indeed the behavior of Ray, but you're right that the mock test doesn't actually prove that this is the behavior of Ray.

> > If we change the "dashboard health check" to use `/api/version` would that be preferable? That way we don't have to introduce logic to deal with the 404 from GetJobInfo and worry about how to test that logic
> 
> Does successfully retrieving `/api/version` from the dashboard indicate that it is ready for job requests? If so, this would be a better solution.

Yes, it does. @kingeasternsun @rueian shall we move the check to use `/api/version` in this PR? I think we can still keep the new `ErrJobInfoNotFound`, it seems useful.

**Comment by kingeasternsun** [comment]
> > > @kevin85421 Hmm, what behavior are you interested in here? Don't the existing RayJob e2e tests cover this?
> > 
> > 
> > I am wondering what the Ray dashboard will return if users look up a non-existent JobID. Will it return a 404? I am not sure whether the mock tests have the same behavior as the Ray dashboard or not.
> 
> Ah I see. Yes, that is indeed the behavior of Ray, but you're right that the mock test doesn't actually prove that this is the behavior of Ray.
> 
> > > If we change the "dashboard health check" to use `/api/version` would that be preferable? That way we don't have to introduce logic to deal with the 404 from GetJobInfo and worry about how to test that logic
> > 
> > 
> > Does successfully retrieving `/api/version` from the dashboard indicate that it is ready for job requests? If so, this would be a better solution.
> 
> Yes, it does. @kingeasternsun @rueian shall we move the check to use `/api/version` in this PR? I think we can still keep the new `ErrJobInfoNotFound`, it seems useful.

I agree with that.

File: ray-operator/controllers/ray/rayjob_controller.go (hunk 2 above)
Comment (kingeasternsun) [reply ray-operator/controllers/ray/rayjob_controller.go]: Thanks, I have changed it by using /api/version to check readiness

**Pull request closed without merging**